File: String/76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int n1=s.size();
        int n2=t.size();
        if(n2>n1) return "";
        vector<int> need(128,0);
        for(char c:t){
            need[c]++;
        }
        int left=0,right=0;
        int required=n2;
        int start=0;
        int minLen=INT_MAX;
        while(right<n1){
          char c=s[right];
          if(need[c]>0) required--;
          need[c]--;
          right++;
          while(required==0){
            if(minLen>right-left){
                minLen=right-left;
                start=left;
            }
            char c=s[left];
            need[c]++;
            if(need[c]>0) required++;
            left++;
          }
            
        }
        return minLen==INT_MAX ? "" : s.substr(start,minLen);
    }
};

File: Linked_List/76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {

        int n1 = t.length();
        int n2 = s.length();

        if(n1>n2) return "";

        int[] freq = new int[128];
      
        for(int i=0;i<n1;i++){

            freq[t.charAt(i)]++;
        }


        int left=0,right=0;

        int required=n1;

        int minlen= Integer.MAX_VALUE;

        int start=0;
       while(right<n2){

        char c = s.charAt(right);

        if(freq[c]>0) required--;

        freq[c]--;

        while(required==0){

            if(right-left<minlen){

                minlen = right-left;

                start=left;
            }

            char ch = s.charAt(left);
                freq[ch]++;
            if(freq[ch]>0) required++;

            left++;
        }

        right++;
       }  

      return minlen==Integer.MAX_VALUE ? "" : s.substring(start,start+minlen+1);    
    }

}